### PR TITLE
simplify and correct the deepspeed example

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?

1. Fixes https://github.com/huggingface/accelerate/issues/1767 and https://github.com/huggingface/accelerate/issues/1773
2. Simplifies the example by removing the redundant saving/loading utils for DeepSpeed as they are supported via `accelerator.save_state()` and `accelerator.load_state()`.

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks!

**Comment by sakshidotvakilsearch** [comment]
Hi can someone explain how these 5 lines of code really work in terms of accelerate + deepspeed?

```
with accelerator.accumulate(model):
          outputs = model(**batch)
          loss = outputs.loss
          accelerator.backward(loss)
          optimizer.step()
          lr_scheduler.step()
          optimizer.zero_grad()
```

line 618-624 in https://github.com/huggingface/accelerate/blob/main/examples/by_feature/deepspeed_with_config_support.py

Looking at the source code, accelerator.backward() will call the deepspeed backward(), which maybe handling some gather-reduce-scatter operations internally (please feel free to elaborate on this too). But post that, the optimizer & scheduler are stepped & optimiser is reset too. Is it a bug that the global step value is not checked before resetting the optimiser such that `global_step % gradient_accumulation_steps == 0`

As far as I understand, deepspeed & accelerate are handling gradient accumulation internally but is the global step tracked by accelerate such that when accelerate.accumulate() is called, it checks whether it is the right step to perform backpass or not? Also, is accumulate triggering the gather operation?

**Comment by sakshidotvakilsearch** [comment]
@pacman100 the completed_steps calculation when training is resumed from a checkpoint that as saved with steps, resumes training from a lot earlier than the step from which it should be resuming. The resume step is supposed to be calculated for epoch while completed_steps is global. so I think this is how it should look like
```
if 'epoch' in training_difference:
    starting_epoch = int(training_difference.replace('epoch_', '')) + 1
    resume_step = None
    completed_steps = starting_epoch * num_update_steps_per_epoch
else:
    resume_step = int(training_difference.replace('step_', ''))
    completed_steps = resume_step
    starting_epoch = resume_step // num_update_steps_per_epoch
    resume_step -= starting_epoch * num_update_steps_per_epoch
```

Your code assigns completed_steps to resume_steps at the end after resume_step calculation

https://github.com/huggingface/accelerate/blob/727d624322c67db66a43c559d8c86414d5ffb537/examples/by_feature/deepspeed_with_config_support.py#L603C20-L603C20
